### ISSUE TITLE
Permission Question: Finding Principals Primary and Secondary...

### ISSUE THREAD

**Issue by 2P2-Anthony** [SharePoint/PnP-Partner-Pack#156]
v2.0

The new Contacts / Principal search is hitting the following URL:

https://<pnp web site>.azurewebsites.net/Persona/SearchPeopleOrGroups?searchText=<name>&searchGroups=false

I can only return one contact successfully, and that is the owner of the Development Subscription we are using. Previously with v1.1 I could search and add anyone in our Company Azure AD Directory. I'm assuming I have the permissions configured incorrectly. However, on triple check, I have them setup as required by the docs:

**MS Graph**
![image](https://cloud.githubusercontent.com/assets/6393228/18862605/1f71d78e-8442-11e6-95dc-c7c827bfed0d.png)

**Azure Active Directory**
![image](https://cloud.githubusercontent.com/assets/6393228/18862613/2b78b71e-8442-11e6-84ad-336f6d0fbfbd.png)

[I don't know why Azure AD shows two delegated permissions on the new portal and only one on the classic portal.]

![image](https://cloud.githubusercontent.com/assets/6393228/18862660/6361404c-8442-11e6-90b9-ecc372d6ee21.png)

**SharePoint Online**
![image](https://cloud.githubusercontent.com/assets/6393228/18862625/353dd2de-8442-11e6-8423-3b1e14e8e7af.png)

No matter what I enter (except the dev account name) I get back the following empty JSON object:

{"Principals":[],"MaxSelectableProfilesNumber":1,"IncludeGroups":false}

Which throws a JS processing error, BTW.

**Comment by PaoloPia** [comment]
Hi,
The behavior that you describe looks weird to me, too. Can you try to download the application manifest, from Azure AD, and see if the "requiredResourceAccess" property looks like the following one?

  "requiredResourceAccess": [
    {
      "resourceAppId": "00000003-0000-0000-c000-000000000000",
      "resourceAccess": [
        {
          "id": "06da0dbc-49e2-44d2-8312-53f166ab848a",
          "type": "Scope"
        },
        {
          "id": "b340eb25-3456-403f-be2f-af7a0d370277",
          "type": "Scope"
        },
        {
          "id": "b4e74841-8e56-480b-be8b-910348b18b4c",
          "type": "Scope"
        }
      ]
    },
    {
      "resourceAppId": "00000002-0000-0000-c000-000000000000",
      "resourceAccess": [
        {
          "id": "311a71cc-e848-46a1-bdf8-97ff7156d8e6",
          "type": "Scope"
        }
      ]
    },
    {
      "resourceAppId": "00000003-0000-0ff1-ce00-000000000000",
      "resourceAccess": [
        {
          "id": "678536fe-1083-478a-9c59-b99265e6b0d3",
          "type": "Role"
        },
        {
          "id": "c8e3537c-ec53-43b9-bed3-b2bd3617ae97",
          "type": "Role"
        }
      ]
    }
  ],

**Comment by 2P2-Anthony** [comment]
Mine has more data, but overlaps your example:

`  "requiredResourceAccess": [
    {
      "resourceAppId": "00000002-0000-0000-c000-000000000000",
      "resourceAccess": [
        {
          "id": "5778995a-e1bf-45b8-affa-663a9f3f4d04",
          "type": "Scope"
        },
        {
          "id": "311a71cc-e848-46a1-bdf8-97ff7156d8e6",
          "type": "Scope"
        }
      ]
    },
    {
      "resourceAppId": "00000003-0000-0000-c000-000000000000",
      "resourceAccess": [
        {
          "id": "b4e74841-8e56-480b-be8b-910348b18b4c",
          "type": "Scope"
        }
      ]
    },
    {
      "resourceAppId": "00000003-0000-0ff1-ce00-000000000000",
      "resourceAccess": [
        {
          "id": "678536fe-1083-478a-9c59-b99265e6b0d3",
          "type": "Role"
        },
        {
          "id": "c8e3537c-ec53-43b9-bed3-b2bd3617ae97",
          "type": "Role"
        }
      ]
    }`

**Comment by 2P2-Anthony** [comment]
@PaoloPia 

I'm not seeing a difference: I copied from the Azure Portal -> Web App -> Edit Manifest

  "requiredResourceAccess": [
    {
      "resourceAppId": "00000002-0000-0000-c000-000000000000",
      "resourceAccess": [
        {
          "id": "311a71cc-e848-46a1-bdf8-97ff7156d8e6",
          "type": "Scope"
        }
      ]
    },
    {
      "resourceAppId": "00000003-0000-0000-c000-000000000000",
      "resourceAccess": [
        {
          "id": "b4e74841-8e56-480b-be8b-910348b18b4c",
          "type": "Scope"
        },
        {
          "id": "b340eb25-3456-403f-be2f-af7a0d370277",
          "type": "Scope"
        },
        {
          "id": "06da0dbc-49e2-44d2-8312-53f166ab848a",
          "type": "Scope"
        }
      ]
    },
    {
      "resourceAppId": "00000003-0000-0ff1-ce00-000000000000",
      "resourceAccess": [
        {
          "id": "678536fe-1083-478a-9c59-b99265e6b0d3",
          "type": "Role"
        },
        {
          "id": "c8e3537c-ec53-43b9-bed3-b2bd3617ae97",
          "type": "Role"
        }
      ]
    }

**Comment by petecuttriss** [comment]
Hi guys, I'm also getting odd behaviour with the Principal selection for v2.0.   

In our case only a select few profiles are returned when you issue a query.  I haven't yet been able to figure out what the difference is between the profiles that are being returned versus those than aren't available.  

With v1.1, I was able to select from any of them.

Regards

**Comment by WorTony** [comment]
I'm also seeing an issue with v2.0 when I try to add Primary and/or Secondary Site Collection Administrator - the webpage displays an error and I'm unable to add any user as a principal

**Comment by WorTony** [comment]
Updating Newtonsoft.json and modifying my web.confing to support the new version fixed it for me

**Comment by awalman** [comment]
I'm unable to add any user as a principal too: 
Exception of type 'Microsoft.Graph.ServiceException' was thrown at OfficeDevPnP.PartnerPack.SiteProvisioning.Components.PeoplePickerHelper.<GetPeoplePickerSearchData>d__1.MoveNext()

I've compared permissions in Azure AD, and using the manifest, and the look the same as those above. 

I'm using latest Newtonsoft.json and December release of PnP - site is an upgrade of v1.1.

**Comment by PaoloPia** [comment]
Hi,
this is the code that searches, using the Microsoft Graph SDK, for users: 

https://github.com/SharePoint/PnP-Partner-Pack/blob/master/OfficeDevPnP.PartnerPack.SiteProvisioning/OfficeDevPnP.PartnerPack.SiteProvisioning/Components/PeoplePickerHelper.cs#L51

As you can see we search for users with DisplayName or UserPrincipalName starting with the search keyword, and we retrieve the following properties: DisplayName, UserPrincipalName, Mail.
Moreover, we keep only those users that have an email address in their profile, because we need it.

I hope this will help.

**Comment by awalman** [comment]
Thanks Paolo. It doesn't seem to matter what we're searching for - even a single letter results in an error. All our accounts have emails, as it's tied to our production Azure AD, in which we have Exchange mailboxes too. I also checked to make sure we haven't turned off Graph access or something like that.

The only difference I can suggest from the original post is that our Azure subscription (in which the web app/storage is hosted) isn't owned by anyone in the target Azure AD (where SharePoint is) - it's under a LiveID and doesn't have a Directory associated with it at all. Tenuous I know, but could this make a difference?

**Comment by PaoloPia** [comment]
Hi,
Having the PnP Partner resources hosted in an Azure subscription separated from the one that holds the Azure AD of the Office 365 tenant is not an issue at all. I usually do the same for my customers. I would try to make the exact same search query using Fiddler (or any other comparable tool) and the same Access Token used by the PnP Partner Pack (you can extract it doins some debugging of the application) and I would see what happens.

**Comment by ypcode** [comment]
Hi,

Just to notice, because I had the same issue, only my user was found in the people picker because it is the only account that has an e-mail address set, (I verified it via the graph explorer).

And it is by design since a site coll admin has to have an e-mail address properly set, the code filters only the "valid" users that have one:
Task<PrincipalsViewModel> MapUsers(IGraphServiceUsersCollectionPage source)

Thus, it is a normal behavior

However, I just set the e-mail address for two accounts in my AD (and in their SharePoint user profiles) but Graph still returns null for the e-mail address, maybe I have to wait for some sync job.

Hope it will help.

Regards,
Yannick
